### PR TITLE
feat(web/desktop): app UI zoom with keyboard shortcuts, Settings control, and desktop persistence

### PR DESCRIPTION
**Pull request by Mirai3103**
#2785

## Why

Ran into the missing zoom support while using OD on a high-DPI display — the app chrome was uncomfortably small and OS-level scaling affected every other app. The standard `Ctrl/Cmd + +/-/0` shortcuts did nothing in the web build, and on desktop the Electron View-menu zoom reset on every restart with no way to set it from within the app.

The pain: users with low vision, high-DPI screens, or non-default viewing distances had no in-app path to scale the UI. WCAG expects content to remain usable at 200% text size; we were below that bar.

## What users will see

- **Web**: `Ctrl/Cmd + =` / `-` / `0` zoom the app interface in and out and reset. The setting persists across reloads.
- **Desktop**: the Electron View-menu shortcuts (`Cmd/Ctrl + +/-/0`) now persist across restarts — zoom no longer resets on quit.
- **Settings → Appearance**: a new "Interface scale" segmented control (80 % – 200 %) appears in both web and desktop. Desktop reads and writes the native Electron zoom factor; web drives the CSS zoom.
- Generated prototype previews and artifact canvases are **not** affected — they counter-scale back to their own size independently.

## Surface area

- [x] **UI** — new "Interface scale" field in Settings → Appearance (`SettingsDialog.tsx`)
- [x] **Keyboard shortcut** — `Ctrl/Cmd + =/-/0` for app-UI zoom (web only; desktop uses existing Electron menu roles)
- [ ] **CLI / env var**
- [ ] **API / contract**
- [ ] **Extension point**
- [x] **i18n keys** — `settings.uiScale`, `settings.uiScaleHint`, `settings.uiScaleReset` added across all 18 locales
- [ ] **New top-level dependency**
- [ ] **Default behavior change**
- [ ] **None**

## Bug fix verification

Not a bug fix PR.

## Validation

```
pnpm --filter @open-design/desktop typecheck   # pass
pnpm --filter @open-design/web test            # 2040/2040 pass
```

Web typecheck has a pre-existing `fetch.preconnect` / bun-types failure on `main` — not introduced by this branch.

### CODE REVIEW

**Comment by lefarcen** [comment]
Hey @Mirai3103! 👋 Thanks for tackling this accessibility gap — the WCAG rationale is solid and the keyboard shortcuts match user expectations.

Before I add @qiongyu1999 as product reviewer, could you add:
- A screenshot showing the new "Interface scale" control in Settings → Appearance
- A quick before (100%) / after (150% or 200%) comparison showing the app chrome at different scales

This helps product evaluate the visual impact without needing to spin up a local build — saves everyone time. The rest of the PR body is clear, just need the visual evidence for the habit-path decision.

Related: I see #2785 (your linked issue) already has `enhancement` + `help wanted` labels, so the **feature intent** is pre-approved. This screenshot request is purely for the **implementation review** — making sure the Settings UI is discoverable and the zoom range feels right.

**Review by PerishCode — COMMENTED** [review]
@Mirai3103 thanks for taking this on — the web/desktop split is sensible, the appearance.ts tests cover the core math + CSS application well, and i18n coverage across all 18 locales is complete. One non-blocking corner case to flag inline around the desktop persistence path; everything else looks good to me.



<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=claude-code · An autonomous AI dev team for your GitHub repos.</sub>

File: apps/desktop/src/main/runtime.ts [hunk 1]
```diff
@@ -1245,6 +1297,27 @@ export async function createDesktopRuntime(options: DesktopRuntimeOptions): Prom
...
+  const persistZoomFactor = (factor: number) => {
+    currentZoomFactor = clampZoomFactor(factor);
+    void writePersistedZoomFactor(currentZoomFactor);
+  };
+  window.webContents.on("did-finish-load", () => {
+    if (!window.isDestroyed()) window.webContents.setZoomFactor(currentZoomFactor);
+  });
+  window.webContents.on("zoom-changed", () => {
+    if (!window.isDestroyed()) persistZoomFactor(window.webContents.getZoomFactor());
+  });
```
Comment (PerishCode) [inline apps/desktop/src/main/runtime.ts:1332]: Programmatic reload (DevTools `Cmd+R`, `location.reload()`, packaged updater rehydration) can revert a menu/keyboard zoom change to the stale persisted value.

The inline comment above acknowledges that `zoom-changed` only fires for `Ctrl+wheel`, so menu accelerators (`zoomIn` / `zoomOut` / `resetZoom` roles call `webContents.setZoomFactor` directly without emitting the event) update the live webContents factor but never advance `currentZoomFactor`. The `close` handler papers over this on quit, but `did-finish-load` is the first listener to fire on any reload and will call `setZoomFactor(currentZoomFactor)` with the stale value — silently snapping the renderer back to the pre-menu-zoom state.

In practice this means: user zooms via View menu to 1.5, then anything that triggers a reload (auto-updater swap, navigation, DevTools refresh) drops them back to 1.0 (or whatever was last persisted) until they close the window.

Suggested fix: capture the live factor before the load resets it, e.g.

```ts
window.webContents.on("will-navigate", () => {
  if (!window.isDestroyed()) {
    currentZoomFactor = clampZoomFactor(window.webContents.getZoomFactor());
  }
});
// or hook into `did-start-loading` for the reload-without-navigation case
```

That way `did-finish-load` re-applies the freshest value, including menu changes that never went through `zoom-changed` or the IPC path. Non-blocking — the symptom only surfaces on reload, which is uncommon in a desktop shell — but worth closing while the surrounding code is fresh.

**Comment by Mirai3103** [comment]
Hey @lefarcen! Here's a screen recording showing the Interface scale control
  in Settings -> Appearance,  with a before/after comparison at different
  zoom levels:

  [https://res.cloudinary.com/dozmrmbbm/video/upload/v1779610172/Screencast_From](https://res.cloudinary.com/dozmrmbbm/video/upload/v1779610172/Screencast_From_2026-05-24_15-06-00_dlqmu9.mp4)
  [_2026-05-24_15-06-00_dlqmu9.mp4](https://res.cloudinary.com/dozmrmbbm/video/upload/v1779610172/Screencast_From_2026-05-24_15-06-00_dlqmu9.mp4)

**Comment by lefarcen** [comment]
Perfect, thanks for the video @Mirai3103! The Settings control placement looks good and the zoom range is clear. I've added @qiongyu1999 for product review — she'll take a look at the visual impact before we move forward. PerishCode's code-level feedback looks non-blocking, so you're in good shape.

**Comment by lefarcen** [comment]
Nice quick iteration @Mirai3103 — the viewport variable approach for the dialog overflow looks solid. Product review is in motion; once @qiongyu1999 weighs in and @PerishCode gives the updated code a thumbs-up, you're good to go.

File: apps/web/src/state/appearance.ts [hunk 2]
```diff
@@ -60,4 +97,32 @@ export function applyAppearanceToDocument({
...
+    // At scale=1 remove the overrides so the default CSS rule wins.
+    if (scale !== 1) {
+      const pct = (100 / scale).toFixed(4);
+      root.style.setProperty('width', `${pct}vw`);
+      root.style.setProperty('height', `${pct}vh`);
+      root.style.setProperty('overflow', 'hidden');
+      // Dialogs and overlays use 100vh/100vw for their size caps. These units
+      // bypass CSS zoom and reference the physical viewport, so at scale>1 a
+      // dialog with max-height:620px overflows (620px*scale > viewport). Expose
+      // zoom-adjusted equivalents that all size-capped elements can consume.
```
Comment (PerishCode) [inline apps/web/src/state/appearance.ts:117]: The new comment says these vars are exposed so "all size-capped elements can consume" them, but several viewport-relative caps in `index.css` weren't switched over. At `uiScale=2` those raw `vh`/`vw`/`dvh` lengths still measure against the un-shrunk physical viewport (since CSS `zoom` on `<html>` doesn't divide them), so the affected elements overflow exactly the way `--vp-*` was added to prevent.

Remaining occurrences in the current `apps/web/src/index.css`:

- L2985 — `.modal-settings { height: 90vh; }`
- L8624 — `width: min(456px, 92vw);`
- L8625 — `height: clamp(480px, 68vh, 640px);`
- L17501 — `max-height: min(90vh, calc(100dvh - 48px));` (sits on the same `.prompt-template-modal` rule whose other `max-height` you migrated on L17500, but this one wins in modern browsers via cascade order, so the var-based fallback never applies)
- L19693 — `width: min(560px, 92vw);`

Either sweep these to the `var(--vp-w, …)` / `var(--vp-h, …)` form (and for the L17500/L17501 pair, fold the modern `100dvh` line into the same expression: `min(90vh, calc(var(--vp-h, 100dvh) - 48px))` or drop the older fallback), or scope the comment here to "the dialog/overlay caps converted in this PR" so it doesn't promise more than the migration covers.

File: apps/web/src/components/SettingsDialog.tsx [hunk 3]
```diff
@@ -5825,10 +5837,80 @@ function AppearanceSection({
...
+  const [factor, setFactor] = useState<number>(DEFAULT_UI_SCALE);
+  useEffect(() => {
+    let alive = true;
+    void getHostZoomFactor().then((value) => {
+      if (alive && typeof value === 'number') setFactor(value);
+    });
+    return () => {
+      alive = false;
+    };
+  }, []);
```
Comment (PerishCode) [inline apps/web/src/components/SettingsDialog.tsx:5913]: `DesktopInterfaceScaleField` reads `getHostZoomFactor()` once on mount (`[]` deps) and never re-fetches, so the segmented control desyncs whenever the user changes zoom from a path that doesn't go through `setHostZoomFactor`:

- View menu (`Cmd/Ctrl + +` / `-` / `0` via the built-in roles)
- Ctrl+wheel, while the dialog is already open

In those cases the live `webContents` factor advances but the local `factor` state stays pinned to the value loaded on dialog open, so the wrong preset stays highlighted (and clicking that preset becomes a no-op IPC round-trip rather than a perceptible UI change). Settings → Appearance is the surface this PR most prominently advertises for scale changes, so the staleness is user-visible.

Same root cause as the existing unresolved thread on `apps/desktop/src/main/runtime.ts` L1308–L1313 (`zoom-changed` only fires for `Ctrl+wheel`, not for menu accelerators). A clean fix on both fronts is to make the main process push zoom transitions to renderers (e.g. an `od:zoom:changed` IPC the renderer can subscribe to via the host bridge), and have the main process emit it from a single `setZoomFactor` choke point — that way (a) menu/IPC/Ctrl+wheel all hit the same notification path, (b) this component can `subscribe` and stay live, and (c) `currentZoomFactor` in `runtime.ts` can't drift between menu changes and a programmatic reload. Cheaper interim: re-query on `visibilitychange` / dialog reopen so it's at least fresh per session.

**Review by PerishCode — COMMENTED** [review]
@Mirai3103 thanks for the follow-up commits adding `--vp-h` / `--vp-w` and the dialog-overflow fix — that pattern is the right shape for keeping the Chromium `zoom` story consistent across modals. Two non-blocking observations on the new code below; the prior unresolved thread on `runtime.ts` (menu-driven zoom + `did-finish-load` reverting on reload) is also still applicable at this head.



<sub>🔁 Powered by <a href="https://github.com/nexu-io/looper">Looper</a> · runner=reviewer · agent=claude-code · An autonomous AI dev team for your GitHub repos.</sub>

**Comment by lefarcen** [comment]
Quick lifecycle check: the latest head `dd47600a` has follow-up commits for the three review threads, and those threads are now resolved. @PerishCode could you re-review the current head when you have a moment? I also looped in @qiongyu1999 through the product-review channel for the visual pass now that the screen recording is attached.

@Mirai3103 nothing extra needed from you right now — this is waiting on reviewer follow-up.
